### PR TITLE
refactor: extract gateway app logic into custom gateway class

### PR DESCRIPTION
**Pull request by alaeddine-13**
Since we aim to implement Custom Gateway in https://github.com/jina-ai/jina/issues/5032, we start by refactoring all logic that creates gateway application (fast api app in http/websocket gateway runtime and grpc.aio.server in grpc gateway).
We try to extract this logic into the `BaseGateway` class and extend this class into 3 subclasses (grpc, http and websockets). These classes are to be exposed to the user in case he wants to customize his gateway.

When implementing this PR we have the following options:
1. let user define everything (app + server) and start/stop server without expecting him to return uvicorn server
* he doesn't expose a server object
* he must implement shutdown and graceful shutdown functionalities (basically everything in gateway runtime). This means we're basicly telling the user to implement gateway runtime.
* user can define any wsgi/asgi server (no limitation on the framework)
* User must implement extra app functionality to preserve compatibility with API (dryrun/health check,...)
2. let user define everything (app + server), start, stop server and he also exposes uvicorn server to us
* we can implement functionalities in gateway runtime for the user (user only cares about defining the app and the server)
* we depend on the server interface (probably only asgi uvicorn server => user can only implement asgi apps. [List of asgi frameworks supported by uvicorn](https://www.uvicorn.org/#asgi-frameworks)
* if there is no flexibility over the server to be exposed, why don't we make the user just implment the asgi app (we allow him to configure the uvicorn server)
* User must implement extra app functionality to preserve compatibility with API (dryrun/health check,...)
3. User only defines the app:
* we expect certain objects for the user per runtime (asgi app to be run on uvicorn server for http/websocket, grpc server for grpc)
* We handle running the server
* User only implements asgi apps
* We relatively can implement extra app functionalities (dryrun/grpc) but but only if we force a framework for the user in http and websockets

In this PR, I decide to adopt option 3 and implement health check/dryrun/reflection features by default only for grpc because we know the app object returned by the user (grpc ai server)

**PS: not implementing health checks might break compatibility with kubernetes deployment**

**Goals:**

- extract gateway app logic into custom gateway class

### CODE REVIEW

File: jina/serve/gateway.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    gateway logic
+
+    :class:`jina.Gateway` as an alias for this class.
+    """
+
+    def __init__(
+        self,
+        streamer: Optional[GatewayStreamer] = None,
+        args: 'argparse.Namespace' = None,
+        logger: 'JinaLogger' = None,
```
Comment (JoanFM) [inline jina/serve/gateway.py:24]: same for logger if it JAML Compatible, it should only have args

File: jina/serve/runtimes/gateway/grpc/__init__.py [hunk 2]
```diff
@@ -20,18 +19,6 @@
...
-    def __init__(
-        self,
-        args: argparse.Namespace,
-        **kwargs,
-    ):
-        """Initialize the runtime
-        :param args: args from CLI
-        :param kwargs: keyword args
-        """
-        self._health_servicer = health.HealthServicer(experimental_non_blocking=True)
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/grpc/__init__.py:32]: why do u need to change this?

**Comment by JoanFM** [comment]
We should only `expose` "start" and "stop" which they need to implement, and our Runtime ensures that they are called properly

File: jina/serve/gateway.py (hunk 1 above)
Comment (alaeddine-13) [reply jina/serve/gateway.py]: it should be jaml compatible now

File: jina/serve/runtimes/gateway/grpc/__init__.py (hunk 2 above)
Comment (alaeddine-13) [reply jina/serve/runtimes/gateway/grpc/__init__.py:32]: no more changed

File: jina/serve/gateway.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        """
+        :param streamer: configured gateway streamer
+        :param args: args
+        :param logger: jina logger object
+        :param kwargs: additional extra keyword arguments to avoid failing when extra params ara passed that are not expected
+        """
+        self.streamer = streamer
+        self.args = args
+        self.logger = logger
+        self.app = self.get_app()
```
[the remark this answers is not in the document]
Comment (alaeddine-13) [reply jina/serve/gateway.py]: now BaseGateway requires server implementation (but not as simple as start and stop, rather, it requires setup, start, teardown and stop)

File: jina/serve/runtimes/gateway/http/gateway.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from jina.helper import extend_rest_interface
+
+from ....gateway import BaseGateway
+from . import get_fastapi_app
+
+
+class HTTPGateway(BaseGateway):
+    """HTTP Gateway implementation"""
+
+    def get_app(self):
```
[the remark this answers is not in the document]
Comment (alaeddine-13) [reply jina/serve/runtimes/gateway/http/gateway.py]: it's no more get_app

File: jina/serve/gateway.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    The base class of all custom Gateways, can be used to build a custom interface to a Jina Flow that supports
+    gateway logic
+
+    :class:`jina.Gateway` as an alias for this class.
+    """
+
+    def __init__(
+        self,
+        streamer: Optional[GatewayStreamer] = None,
```
[the remark this answers is not in the document]
Comment (alaeddine-13) [reply jina/serve/gateway.py]: it is no exposed rn

File: jina/serve/gateway.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        :param name: Gateway pod name
+        :param kwargs: additional extra keyword arguments to avoid failing when extra params ara passed that are not expected
+        """
+        self.streamer = None
+        self.name = name
+        # TODO: original implementation also passes args, maybe move this to a setter/initializer func
+        self.logger = JinaLogger(self.name)
+
+    def _set_streamer(
```
Comment (JoanFM) [inline jina/serve/gateway.py:37]: why is `_set_streamer` not called in the `Init`?

**Review by JoanFM — COMMENTED** [review]
I like it in general, I just want to make sure `_set_streamer` is called in `__init__` (RAII pattern). 

I am not fan of the `should_exit` property but let's keep it for now

**Review by JoanFM — APPROVED** [review]

**Pull request merged**